CRM/Contribute/BAO/Contribution.php

```diff
@@ -176,7 +179,32 @@ public static function add(&$params, $ids = array()) {
     if ($contributionID && $setPrevContribution) {
       $params['prevContribution'] = self::getValues(array('id' => $contributionID), CRM_Core_DAO::$_nullArray, CRM_Core_DAO::$_nullArray);
     }
-
+    // CRM-18573
+    if (!empty($params['contribution_status_id'])) {
+      if ($isEmptyPaymentMethod) {
+        if (!in_array($contributionStatus[$params['contribution_status_id']],
+          array('Completed', 'Partially paid', 'Pending refund'))
+        ) {
+          unset($params['payment_instrument_id']);
+        }
+        elseif (empty($params['payment_instrument_id']) && !empty($params['prevContribution'])
+          && empty($params['prevContribution']->payment_instrument_id)
+        ) {
+          $params['payment_instrument_id'] = key(CRM_Core_OptionGroup::values('payment_instrument', FALSE, FALSE, FALSE, 'AND is_default = 1'));
+        }
+      }
+      if (empty($params['receive_date']) && !(!empty($params['prevContribution']) && !empty($params['prevContribution']->receive_date))
+        && !in_array($contributionStatus[$params['contribution_status_id']],
+          array('Failed', 'Pending')
+        )
+      ) {
+        $params['receive_date'] = date('YmdHis');
+      }
+    }
+    $errorMessage = CRM_Contribute_BAO_Contribution::checkPaymentInstrument($params);
+    if ($errorMessage) {
```

Comment:
checkPaymentInstrument() is also called at form level in formrule method. So just raised an exception in BAO